Cql/Cql.CqlToElm/TypeConverter.cs

```diff
@@ -0,0 +1,426 @@
+
+            Expression FromList(Expression list, TypeSpecifier to) =>
+                As(new SingletonFrom { operand = list }.WithResultType(to), to);
+
+            Expression FromInterval(Expression interval, TypeSpecifier to) =>
+                As(new Start { operand = interval }.WithResultType(to), to);
+
+            Expression ToList(Expression expression, ListTypeSpecifier to) =>
+                new ToList { operand = As(expression, to.elementType) }.WithResultType(to);
+        }
+
+        /// <summary>
+        /// Computes the cost of conversion according to the specification's conversion precendence rules.
+        /// </summary>
+        /// <seealso href="https://cql.hl7.org/03-developersguide.html#conversion-precedence"/>
+        public ConversionCost GetConversionCost(TypeSpecifier from, TypeSpecifier to)
+        {
+            if (IsExactMatch(from, to))
+                return ConversionCost.ExactMatch;
+            else if (IsSubtype(from, to))
+                return ConversionCost.Subtype;
+            else if (IsCompatible(from, to))
+                return ConversionCost.Compatible;
+            else if (CanBeCast(from, to))
+                return ConversionCost.Cast;
+            else if (HasImplicitConversion(from, to))
+            {
+                if (IsSimpleType(to))
+                    return ConversionCost.ImplicitToSimpleType;
+                else if (IsClassType(to))
+                    return ConversionCost.ImplicitToClassType;
+                else return ConversionCost.Incompatible;
+            }
+            else if ((Options.EnableIntervalPromotion ?? false)
+                && to is IntervalTypeSpecifier promotableInterval
+                && CanBePromoted(from, promotableInterval))
+                return ConversionCost.IntervalPromotion;
+            else if ((Options.EnableListDemotion ?? false)
+                && from is ListTypeSpecifier listType && CanBeDemoted(listType, to))
+                return ConversionCost.ListDemotion;
+            else if ((Options.EnableIntervalDemotion ?? false)
+                && from is IntervalTypeSpecifier demotableInterval && CanBeDemoted(demotableInterval, to))
+                return ConversionCost.IntervalDemotion;
+            else if ((Options.EnableListPromotion ?? false)
+                && to is ListTypeSpecifier promotableListType && CanBePromoted(from, promotableListType))
+                return ConversionCost.ListPromotion;
+            else
+                return ConversionCost.Incompatible;
+        }
+
+        internal string IncompatibilityMessage(TypeSpecifier from, TypeSpecifier to) =>
+            $"Expression of type '{from}' cannot be cast as a value of type '{to}'.";
+
+        // All types exactly match Any
+        internal virtual bool IsExactMatch(TypeSpecifier from, TypeSpecifier to) =>
+            from == to;
+
+        internal virtual bool IsSubtype(TypeSpecifier from, TypeSpecifier to)
```

Comment:
The functionality in this class works and passes all tests.  The implementation in `ModelProvider` does not.

We need to review the usage in general around extension methods.  I have a distaste for this pattern:

`from.IsSubtypeOf(to, ModelProvider);`

If we could have made it:

`from.IsSubtypeOf(to)`

Then that is very good usage.  

But if the core functionality of this method has to be implemented through `ModelProvider`, then I find it much more preferable to make the call pattern like this:

`ModelProvider.IsSubType(from, to)`

Reply:
Here's an example test that `ModelProvider`'s `IsSubType` chokes on:
```
        [TestMethod]
        public void IntervalIncludesNull()
        {
            var arguments = new Expression[] { Interval(Integer(1), Integer(10)), Null };
            var result = InvocationBuilder.MatchSignature(SystemLibrary.Contains, arguments);
            result.Compatible.Should().BeTrue();
            result.Arguments[0].Result.Should().HaveType(SystemTypes.IntegerType.ToIntervalType());
            result.Arguments[1].Result.Should().HaveType(SystemTypes.IntegerType);
         }
```
This test matches how the cql-to-elm reference engine produces ELM for a CQL expression of:
`Integer[1,10] contains null`

When using the `ModelProvider`'s implementation, this code gets hit:
```
            else if (subType is Elm.IntervalTypeSpecifier intervalSub && superType is Elm.IntervalTypeSpecifier intervalSuper)
            {
                // covariance on intervals
                return intervalSub.pointType.IsSubtypeOf(intervalSuper.pointType, provider);
            }
```

The current algorithm for inferring generics is to see which inference is cheaper, so we try inferring both `Integer` and `Any` for `Contains(Interval<T>, T)`.

When we infer `T` to be `Any`, we now have a subtype relationsip where `Interval<Integer>` is a subtype of `Interval<Any>` (precedence = 2).  This is cheaper than treating the `null` to be an `Integer` (compatible, precedence = 3), so the conversion precedence requires we pick the cheaper match and we change the first argument, `Interval[1,10]`, to be an `Interval<Any>`.

At the same time, this is true:
`Interval[1,10] is Any`.

The `ModelProvider`'s implementation of `IsSubType` is, I believe, correct.  You can find specification language that justifies every line of code in that method.

You might think the solution should be that we only attempt to 

I have been tearing my hair out the last week trying to create a set of rules that both universally follow the specification but also pass all the tests.  That is impossible.

I have considered that maybe what we should be doing is only inferring the generic on the first argument, and if any acceptable match is found, use that.  But the following expressions are all legal in CQL:

```
define Contains1: Interval[1, 10] contains null
define Contains2: null contains 5
define Contains3: Interval[null, null] contains (5 as Any)
```
`Contains2` infers the `null` operand to be an `Interval<Integer>`.  If the reference implementation used the approach of accepting the first matching generic inference instead of the overall cheapest, it would infer T to be `Any` and then cast the `5` to `Any`, resulting in the legal expression `Contains3`.

Reply:
We need it to look at all arguments, otherwise the Add would not work either -   `Add(4.0,4)` would select `Add(double,double)` as the overload, which is fine, but `Add(4,4.0)` would select the `Add(integer,integer)` overload, which would fail. That was the original reason I had to consider all arguments for assignment of generic parameters.